packages/flutter/lib/src/widgets/widget_inspector.dart

```diff
@@ -0,0 +1,444 @@
+    double x;
+    if (normalizedTargetX > size.width - _kScreenEdgeMargin - childSize.width) {
+      // Make room for content to the right.
+      x = size.width - _kScreenEdgeMargin - childSize.width;
+    } else {
+      x = normalizedTargetX;
+    }
+    _childPosition = new Offset(x, y);
+  }
+
+  final Offset target;
+  final double verticalOffset;
+  final bool preferBelow;
+  final Size size;
+  final Size childSize;
+
+  Offset _childPosition;
+  bool _tooltipBelow;
+
+  bool get tooltipBelow => _tooltipBelow;
+  Offset get childPosition => _childPosition;
+}
+
+// FOR DISCUSSION: this is clearly not quite right: should I even even bother
+// listening for changes? This code is copied from semantics_debugger where
+// it made sense.
+class _WidgetClient extends ChangeNotifier {
+  _WidgetClient(PipelineOwner pipelineOwner) {
+    _semanticsHandle =
+        pipelineOwner.ensureSemantics(listener: _didUpdateSemantics);
+  }
+
+  SemanticsHandle _semanticsHandle;
+
+  @override
+  void dispose() {
+    _semanticsHandle.dispose();
+    _semanticsHandle = null;
+    super.dispose();
+  }
+
+  int generation = 0;
+
+  void _didUpdateSemantics() {
+    generation += 1;
+    notifyListeners();
+  }
+}
+
+const TextStyle _messageStyle = const TextStyle(
+    color: const Color(0xFFFFFFFF), fontSize: 10.0, height: 1.2);
+final int _kMaxTooltipLines = 5;
+const Color _kTooltipBackgroundColor = const Color.fromARGB(230, 60, 60, 60);
+
+void _paintDescription(Canvas canvas, RenderBox render, Offset target,
+    double verticalOffset, Size size, Rect targetRect) {
+  // TODO(jacobr): craft a better description message.
+  final String message = '$render\n\n'
+      '${render.debugCreator}';
+  canvas.save();
+  final TextPainter textPainter = new TextPainter()
+    ..maxLines = _kMaxTooltipLines
+    ..ellipsis = '...'
+    ..text = new TextSpan(style: _messageStyle, text: message)
+    ..layout(
+        maxWidth: size.width - 2 * (_kScreenEdgeMargin + _kTooltipPadding));
+
+  final Size tooltipSize =
+      textPainter.size + const Offset(_kTooltipPadding * 2, _kTooltipPadding * 2);
+  final _TooltipPositionCalculator calc = new _TooltipPositionCalculator(
+      target: target,
+      verticalOffset: verticalOffset,
+      preferBelow: false,
+      size: size,
+      childSize: tooltipSize);
+  final Offset tipOffset = calc.childPosition;
+
+  final Paint tooltipBackground = new Paint()
+    ..style = PaintingStyle.fill
+    ..color = _kTooltipBackgroundColor;
+  canvas.drawRect(
+      new Rect.fromPoints(tipOffset,
+          tipOffset.translate(tooltipSize.width, tooltipSize.height)),
+      tooltipBackground);
+
+  double wedgeY = tipOffset.dy;
+  if (!calc.tooltipBelow) {
+    wedgeY += tooltipSize.height;
+  }
+  final double wedgeSize = _kTooltipPadding * 2;
+  double wedgeX = math.max(tipOffset.dx, target.dx) + wedgeSize * 2;
+  wedgeX = math.min(wedgeX, tipOffset.dx + tooltipSize.width - wedgeSize * 2);
+  canvas.drawPath(
```

Comment:
This is to render a chat bubble style tooltip like chrome devtools. Obviously if we decide this is a good UI to make it clear what object the tip is associated with we should polish this code as the border for the bubble isn't right.